## Function and Files

### Function

<br>

**함수(function)**는 **특정 작업을 수행하는 명령어들의 모임**에 이름을 붙인 것이다
- 함수는 작업에 필요한 데이터를 전달받을 수 있으며, 작업이 완료 된 후에는 작업의 결과를 호출자에게 반환할 수 있다
- 자주 사용되는 부분 문제를 함수로 작성하면 코드를 **반복 작성할 필요가 없기 때문에 편리**하며, 호출하여 재사용하면 된다
- 파이썬에서 지원하는 세 종류의 함수
    - 내장(Built-in) 함수 : 파이썬에서 제공하는 함수
        - 파이썬 설치 후 사용 가능(print(  ), input(  ), len(  ), type(  ) .....)
    - 라이브러리 패키지 : 파이썬에서 제공하는 모듈(import 문)
        - 해당 모듈을 프로그램에 포함한(loading)후에 사용할 수 있다
    - 사용자 정의(User-defined) 함수
        - 사용자가 자신의 필요에 따라 특정 기능의 함수를 직접 작성해서 쓴다

In [47]:
def plus_one(n) :
    result = n + 1
    return result

6

In [48]:
print('123')

123


In [1]:
print('this is function') # print function

this is function


In [2]:
round(3.141592, 2) # round function

3.14

In [3]:
str(1) # str function

'1'

In [4]:
int('100') # int function

100

In [5]:
list((1,2,3,4,5)) # list function

[1, 2, 3, 4, 5]

In [49]:
tuple([1,2]) # tuple

(1, 2)

**type 함수**
- 대상에 대한 type을 출력한다. 대상의 type을 알고 싶으면 type을 호출해보면 알 수 있다

In [6]:
type(1) # type function

int

In [7]:
type('FINDA')

str

In [8]:
type(dir)

builtin_function_or_method

**id 함수**
- 인자로 받은 parameter의 식별값을 구한다. python interpreter는 모든 대상체에 대해서 식별값을 부여한다. id 함수는 이 식별값을 구해준다.

In [9]:
id(3.14)

140680885775568

In [10]:
id('FINDA')

140680313076784

**sorted 함수**
- sorted 함수는 반복 가능한 데이터를 입력 받아서 정렬된 리스트로 반환한다
- key는 정렬한 기준을 넘겨준다. 만약 정렬기준이 없으면 기본 정렬 기준을 이용한다
- reverse는 역순 정렬 시킬 것인지의 여부를 알려주는 인자이다

In [55]:
numbers = [2,3,1,5,4]
sorted_numbers = sorted(numbers)
sorted_numbers # 오름차순 정렬

[1, 2, 3, 4, 5]

In [12]:
reverse_sorted_numbers = sorted(numbers, reverse = True)
reverse_sorted_numbers # 내림차순 정렬

[5, 4, 3, 2, 1]

In [56]:
numbers = [2,3,1,5,4]
sorted(numbers)

[1, 2, 3, 4, 5]

In [58]:
numbers.sort()

In [59]:
numbers # 함수의 결과가 저장되지는 않는다

[1, 2, 3, 4, 5]

In [60]:
colors = ['red','green','blue','yellow']

In [64]:
sorted_colors = sorted(colors, reverse = True) # 사전 알파벳 순으로 정렬된다
sorted_colors

['yellow', 'red', 'green', 'blue']

In [62]:
len_sorted_colors = sorted(colors, key = len) # argument key로 정렬 기준 지정이 가능하다
len_sorted_colors

['red', 'blue', 'green', 'yellow']

In [63]:
reverse_len_sorted_colors = sorted(colors, key = len, reverse = True)
reverse_len_sorted_colors

['yellow', 'green', 'blue', 'red']

#### Library(Package) Function
- python에서는 함수들을 기능별로 패키지 혹은 모듈에 포함시켜서 제공한다
- 모듈이란 python 코드를 포함하고 있는 파일을 말한다
- 모듈들을 묶어서 package라고 한다
- python을 설치하면 많은 Library(package)와 모듈들이 설치되고 각 라이브러리 및 모듈마다 많은 수의 함수들이 포함되어 있다
- 기본 라이브러리에서 제공되지 않는 것은 추가의 라이브러리를 설치해서 사용할 수 있다

#### import 예약어
- 라이브러리나 모듈을 포함시킨다
- import *Library/Module*
- import *Library/Module* as *Lib/Mod*
    - 라이브러리나 모듈을 축약된 이름으로 바꿔서 포함시킨다. 이 경우 축약이름으로 사용해야 한다
- from *Library/Module* import *func*
    - 라이브러리나 모듈로부터 몇몇 대상체만을 포함시킨다
    - 이 경우 대상체의 사용은 대상체명을 이용해서 사용한다
- from *Library/Module* import *
    - 라이브러리나 모듈 내에 포함된 모든 대상체를 import 한다
    - 이 경우 대상체의 사용은 대상체의 이름을 이용해서 사용한다
    - 다만, 함수간 충돌이 일어날 가능성이 크기 때문에 사용하지 않는것이 좋다

In [65]:
dir(math)

['__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'acos',
 'acosh',
 'asin',
 'asinh',
 'atan',
 'atan2',
 'atanh',
 'ceil',
 'comb',
 'copysign',
 'cos',
 'cosh',
 'degrees',
 'dist',
 'e',
 'erf',
 'erfc',
 'exp',
 'expm1',
 'fabs',
 'factorial',
 'floor',
 'fmod',
 'frexp',
 'fsum',
 'gamma',
 'gcd',
 'hypot',
 'inf',
 'isclose',
 'isfinite',
 'isinf',
 'isnan',
 'isqrt',
 'lcm',
 'ldexp',
 'lgamma',
 'log',
 'log10',
 'log1p',
 'log2',
 'modf',
 'nan',
 'nextafter',
 'perm',
 'pi',
 'pow',
 'prod',
 'radians',
 'remainder',
 'sin',
 'sinh',
 'sqrt',
 'tan',
 'tanh',
 'tau',
 'trunc',
 'ulp']

In [18]:
import math # module import (library import)
math.sqrt(2) # module function

1.4142135623730951

In [66]:
import math as m
m.sqrt(2)

1.4142135623730951

In [19]:
from math import sin, log, pi, e
sin(pi)

1.2246467991473532e-16

In [20]:
log(e)

1.0

#### os Library
- os 라이브러리는 운영체제 관련 속성, 함수, 및 클래스들을 포함하고 있다

#### sys Library
- sys 라이브러리는 파이썬 시스템 관련 속성, 함수, 및 클래스들을 포함하고 있다

In [21]:
import os

dir(os)[:5] # dir함수를 쓰면 라이브러리의 속성을 볼 수 있다

['CLD_CONTINUED', 'CLD_DUMPED', 'CLD_EXITED', 'CLD_KILLED', 'CLD_STOPPED']

In [22]:
import sys

dir(sys)[:5]

['__breakpointhook__',
 '__displayhook__',
 '__doc__',
 '__excepthook__',
 '__interactivehook__']

#### 사용자 정의 함수

- 함수를 작성하는 이유는 크게 두 가지로 볼 수 있다.
    - <span style = 'color : #132c6f'><b>특정의 기능을 수행하는 코드들을 하나의 묶음</b></span>으로 사용하기 위해 함수를 사용한다.
    - <span style = 'color : #132c6f'><b>동일한 기능을 수행하는 코드들의 재사용성을 높이고 코드의 통일된 관리</b></span>를 하기 위해 함수를 작성한다.

<br>
<pre><b>def</b> function_name(parameters) :
<pre><i>"""docstring"""</i>
statements
<b>return</b> ret_values</pre></pre>
    
- 함수 작성 (함수 정의)
- def:함수의 시작을 알린다
    - 함수 이름(function_name) : identifier(변수) 규칙대로 이름 정의
- parameters(매개변수)
    1. 함수의 입력 값(콤마로 분리), 필요하지 않으면 괄호만 표시한다
    2. 함수 호출시 전달하는 데이터는 함수의 매개변수로 전달된다
    3. 함수 호출시 전달하는 데이터를 인수(arguments)라 한다
- """docstring""" : 주석 (함수 설명), 생략도 가능하다
- return: 실행 결과를 호출한 코드로 반환, 반환값이 없으면 생략한다

#### 함수 호출하기

- 프로그램에서 함수를 사용하려면 함수를 호출(call)하여야 한다
- 함수 호출시 데이터(값,정보)를 지정하여 함수에 전달할 수 있다
    - 이 데이터를 **인수(argument)**라고 함 (위치에 따른 인수 전달 방식)
- 인수와 매개변수는 함수 호출 시에 데이터를 주고받기 위하여 필요 하며, 함수가 호출될 때마다 인수는 함수의 매개변수로 전달된다
    - **인수(argument)** : 호출 프로그램에 의하여 함수에 전달되는 값
    - **매개 변수(parameter)** : 함수에서 인수를 전달받는 변수
- 한번 정의된 함수는 필요시 반복해서 호출할 수 있다 (단, shell을 한번 종료하면 다시 실행시켜야 한다)
- 함수가 호출되는 시점에서 실행의 흐름은 함수로 넘어가며, 함수 종료 후에는 호출된 곳으로 돌아온다
- 파이썬 코드에서는 **들여쓰기 하지 않은 첫 명령어부터 실행**되며 이를 main 함수로 간주한다

In [68]:
def null() :
    print('null function')
    return 0
    
null() # function calling
a = null() # function calling
print(a) # return값 저장 후 출력

null function
null function
0


In [24]:
def test_function(arg) :
    """This is docstring""" # 함수의 주석
    
    print('function calling')
    temp = str(arg) * 2 # 함수 내에 다른 함수 호출 가능
    # temp를 지역변수 (local variable)이라고 함
    return temp

In [69]:
test = test_function(1) # function calling
print(test) # test에 return값이 저장됨

function calling
11


In [70]:
help(test_function) # Docstring 확인 가능

Help on function test_function in module __main__:

test_function(arg)
    This is docstring



### Exercise 1.
소수 판별 알고리즘을 함수로 만들어 보자

In [73]:
def prime_number_1(x) :
    plag = True # boolean data
    for i in range(2, x) : 
        if x % i == 0 : plag = False # 2 이상의 수에서 만약 나누어 떨어지면 소수가 아니다

    if plag == True : print('Prime Number')
    else : print('Not Prime Number')

In [27]:
def prime_number(num) :
    plag = True
    for i in range(2, num) : 
        if num % i == 0 : plag = False 

    if plag == True : print('Prime Number')
    else : print('Not Prime Number')

In [74]:
prime_number_1(100)

Not Prime Number


In [75]:
prime_number_1(7)

Prime Number


#### argument가 두 개인 경우

In [76]:
def print_sum(n1, n2) :
    result = n1 + n2
    print('The result is %i' % int(result))
    return result

In [31]:
print_sum(3,5)

The result is 8


8

#### 두개의 반환값을 return하는 함수

In [32]:
def circle_area_circumference(radius, pi) :
    area = pi * (radius ** 2)
    circumference = 2 * pi * radius
    return area, circumference

In [33]:
circle_area_circumference(3, 3.14) # tuple의 형태로 출력된다

(28.26, 18.84)

In [34]:
result = circle_area_circumference(3, 3.14)
print('반지름 :', 3, ' 면적과 둘레 :', result)
res1, res2 = circle_area_circumference(3, 3.1415)
print('반지름 :', 3, '면적 :', res1, '둘레 :', res2)

반지름 : 3  면적과 둘레 : (28.26, 18.84)
반지름 : 3 면적 : 28.273500000000002 둘레 : 18.849


python에서는 함수의 결과로 여러 개의 값을 반환할 수 있다. return 다음에 적어주며, 반환값을 tuple의 형태로 돌려준다.

tuple을 그대로 이용할 수도 있고, **unpacking**해서 각 요소를 구해 이용 가능하다. (split해서 사용가능하다)

### Parameter
- General parameter, Default parameter, Keyword parameter, Variable parameter
- 일반 인자는 호출시에 위치가 매우 중요하다
- 함수 호출 시에 넣어준 인자 순서대로 인자값이 함수에 전달된다. 이는 가장 많이 사용되는 함수 인자 전달 방식이다
- python에서는 일반 인자 이외에도 <span style = 'color : #134f2c'><b>특별한 인자 전달 방식</b></span>을 제공한다
    - 기본인자, 키워드 인자, 가변인자, 정의되지 않은 인자

#### Default parameter
- 함수를 호출할 때 <span style = 'color : #132c6f'><b>인자의 값을 설정해 주지 않아도 기본 값이 할당</b></span>되도록 하는 인자이다
- 기본 인자는 일반 인자 뒤에 위치한다

In [77]:
def future_to_present(future, year, interest = 0.03) : # default value 0.03 지정
    present = future / (1 + interest) ** year 
    return present

print('미래가치 : %d, 만기 : %d년, 금리 : 0.03, 현재가치 : %.2f' % (1000, 5, future_to_present(1000, 5)))
print('미래가치 : %d, 만기 : %d년, 금리 : 0.05, 현재가치 : %.4f' % (1000, 5, future_to_present(1000, 5, 0.05)))

미래가치 : 1000, 만기 : 5년, 금리 : 0.03, 현재가치 : 862.61
미래가치 : 1000, 만기 : 5년, 금리 : 0.05, 현재가치 : 783.5262


#### Keyword parameter
- 함수에 인자 값을 전달할 때 <span style = 'color : #132c6f'><b>인자의 이름으로 인자 값을 전달</b></span>하는 방법이다
- 키워드 인자는 일반 인자 뒤에 나온다
- <span style = 'color : #760c0c'>키워드 인자 이후에는 순서에 의한 인자 매칭을 할 수 없다</span>

In [80]:
def future_to_present(future, year, interest = 0.03) : # default value 0.03 지정
    present = future / (1 + interest) ** year 
    return present

print('미래가치 : %d, 만기 : %d년, 금리 : 0.03, 현재가치 : %.2f' 
      % (1000, 5, future_to_present(1000, year = 5)))
print('미래가치 : %d, 만기 : %d년, 금리 : 0.05, 현재가치 : %.4f' 
      % (1000, 5, future_to_present(year = 5, interest = 0.05, future = 1000)))
# 키워드 인자를 쓰기 시작하면, 더이상 순서에 의한 매칭을 기대할 수 없다

미래가치 : 1000, 만기 : 5년, 금리 : 0.03, 현재가치 : 862.61
미래가치 : 1000, 만기 : 5년, 금리 : 0.05, 현재가치 : 783.5262


#### Variable parameter
- 개수에 제한이 없다
- 함수를 정의할 때, <span style = 'color : #132c6f'><b>함수 인자 앞에 *을 붙이면 정해지지 않은 수의 인자를 받겠다는 뜻</b></span>이다
- 가변 인자 리스트는 입력 받은 인자를 tuple에 저장한다

In [37]:
def circle_area(pi, *radiuses) :
    areas = []
    for radius in radiuses :
        area = pi * (radius ** 2)
        areas.append(area)
    return areas

print(circle_area(3.14, 3, 4, 5))
print(circle_area(3.14, 6, 7))

[28.26, 50.24, 78.5]
[113.04, 153.86]


In [38]:
def circle_area(pi, *radiuses) :
    areas = []
    
    for radius in radiuses :
        area = pi * (radius ** 2)
        areas.append(area)
        
    return areas

print(circle_area(3.14, *(3,4,5))) # tuple앞에 *를 붙이면 가변 인자의 형태로 넘어간다
print(circle_area(3.14, *(6,7)))

[28.26, 50.24, 78.5]
[113.04, 153.86]


#### Undefined Parameter
- python에서는 함수에 정의되지 않은 인자들을 함수에 전달할 수 있다
- **정의되지 않은 인자는 사전 형식으로 전달되며**, 함수에서 정의되지 않은 인자를 받을 때는 인자 앞에 ** 를 붙여서 받는다
- 일반인자, 가변인자, 정의되지 않은 인자가 있을 경우 함수에서는 <span style = 'color : #760c0c'>반드시 일반인자, 가변인자, 정의되지 않은 인자 순으로 </span>인자들을 전달하고 받아야 한다

In [39]:
def circle_area(radius, *pi, **info) :
    for item in pi :
        area = item * (radius ** 2)
        print('반지름 : ', radius, 'PI : ', item, '면적 : ', round(area, 2))
    for key in info : print(key, ':', info[key])
        
circle_area(3, 3.14, 3.1415, line_color = '파랑', area_color = '노랑')
print()
circle_area(5, 3.14, 3.1415, polygon_name = '원', value = '면적')

반지름 :  3 PI :  3.14 면적 :  28.26
반지름 :  3 PI :  3.1415 면적 :  28.27
line_color : 파랑
area_color : 노랑

반지름 :  5 PI :  3.14 면적 :  78.5
반지름 :  5 PI :  3.1415 면적 :  78.54
polygon_name : 원
value : 면적


### Exercise 2.
현재가치를 미래가치로 만드는 함수를 만들어 보자

In [81]:
def present_to_future(present, year, interest) :
    future = present * (1 + interest) ** year
    return future

In [82]:
present_to_future(1000, 5, 0.04)

1216.6529024000001

### File I/O
- 키보드 입력을 표준 입력, 모니터 출력을 표준 출력이라 한다
- input( ) : 키보드로부터 데이터를 입력 받는 함수
- print( ) : 모니터로 결과를 출력하는 함수(옵션을 변경하여 파일에 출력 가능)
     - 하지만, 입력 데이터의 양이 많아지면 손으로 직접 입력하기 어렵고, 출력 데이터의 양도 많아져 파일에 보관해야 할 필요성도 생김
     
<br>

- 파일 입출력은 다음과 같은 세 단계를 통하여 이루어짐
    1. 파일 열기 (File Open)
    2. 파일에서 읽거나 또는 파일에 쓰기
    3. 파일 닫기 (File Close)

#### 파일 열기

<pre>
fp_r = open("in.txt", 'r‘) # <i>read mode</i> 
fp_w = open("out.txt", 'w‘) # <i>write mode</i> </pre>

- fp_r, fp_w : 파일 객체를 나타내는 변수
- in.txt, out.txt : 사용할 파일 이름
- File Mode : 파일을 어떤 용도로 사용할지 결정
    - 'r' : 읽기 전용. 존재하지 않는 파일이면 에러가 발생하므로 주의한다
    - 'w' : 쓰기 전용. 파일이 없으면 빈 파일을 새로 만들고, 기존 있던 파일이면 내용을 모두 지우고 파일을 쓰기 전용으로 open한다
    - 'a' : append의 약자이며, 기존 파일에 덧붙여서 이어서 쓰기 작업을 한다
- script 파일과 같은 폴더에 저장되어 있는 파일을 읽거나 쓰는 경우 는 파일 이름만 명시한다. 다른 폴더에 있는 파일인 경우는 전체 경로를 이름과 함께 명시해야 한다

#### 파일 읽기

- open() 함수에서 읽기 모드로 반환 받은 파일 객체를 가리키는 변수를 fp_r 이라고 가정한다
- **fp_r.read( )** : 파일 전체를 하나의 문자열로 읽는다
- **fp_r.readlines( )** : 파일을 줄 단위로 읽어 각 줄을 문자열 형식으로 저장하고, 이들 전체 문자열을 원소로 하는 리스트를 반환한다
- **fp_r.readline( )** : 한 줄을 문자열 형식으로 읽어 이를 반환한다
- 파일 전체를 한 줄씩 읽어 처리하는 방법 (for 반복문 사용하면 편리하다)

<pre>
<b>for</b> aLine <b>in</b> fp_r : # 변수 aLine은 fp_r에서 한 줄씩 읽어 들인 문자열 
<pre>statements # 읽어들인 줄에 처리할 명령어들</pre></pre>

#### 파일 쓰기
- open() 함수에서 쓰기 모드로 반환 받은 파일 객체를 가리키는 변 수를 fp_w 이라고 가정한한다
- **fp_w.write(str)** : 인자로 받은 문자열 데이터 하나를 파일에 쓴다
    - 줄바꿈 문자가 자동으로 삽입되지 않는다
    - 줄바꿈이 필요한 경우, 문자열 인자의 마지막에 줄바꿈 기호를 추가해야 한다
- **print(\*objects, file = fp_w)**
    - 화면 대신 파일에 쓰기 위해서는 file 인수값 변경이 필요하다
    - print() 함수는 디폴트로 줄바꿈이 발생한다
    - 줄바꿈을 하지 않으려면, end 인수값을 변경해야 한다

#### 파일 닫기
파일을 열었다면 **반드시 파일을 닫아야 한다**

<pre>
fp_r = open("in.txt", 'r‘) # read mode fp_w = open("out.txt", 'w‘) # write mode 
....
....
<b>fp_r.close()</b>
<b>fp_w.close()</b></pre>


- 파일이 열려있는 상태에서는 파일을 지우거나 이름을 바꿀 수 없다
- Python shell을 종료하거나, restart하거나 또는 close() method를 호출 하여야 파일이 닫혀, 파일 지우기 또는 이름 바꾸기 등을 수행할 수 있다
- 따라서, 파일 작업을 마치면 반드시 파일을 닫도록 해야 한다

In [83]:
fp = open('information.txt', 'wt', encoding = 'utf-8') # file writing
fp.write('3\n')
fp.write('3.14\n')
fp.write('"3.14"\n')
fp.close() # 파일 닫기

In [84]:
with open('information.txt', 'rt') as fp : # file open
    print(fp.read()) # 읽기 모드이면 파일을 닫을 필요는 없다 (참조만 했기 때문)

3
3.14
"3.14"



In [85]:
fp = open('information.txt', 'rt', encoding = 'utf-8')
for s in fp :
    print(s)

3

3.14

"3.14"



### Exercise 3.
아래의 내용을 파일에 입력하고 직접 읽어 보자

<pre>
3.141592
FIND-A
This is Python
</pre>

In [86]:
f = open('test.txt', 'wt', encoding = 'utf-8') # cp949
f.write('3.141592\n')
f.write('FIND-A\n')
f.write('This is Python\n')
f.close()

In [87]:
with open('test.txt', 'rt') as f :
    print(f.read())

3.141592
FIND-A
This is Python



### Assignment 1.
**Gordon의 기업 성장 모형**과 **자본자산가치결정모형(CAPM)**에 따르면 투자자가 기업에 요구하는 요구 수익률은 다음과 같이 정의됩니다.

$$k = r_f + \beta(E(r_m) - r_f)$$

여기서 $k$는 요구수익률, $r_f$는 무위험이자율, $r_m$는 기업 주식의 수익률, $\beta$는 기업의 시장 프리미엄을 의미합니다

- 위 산식을 함수로 작성하세요

#### Assignment 1-1.

- 2023년 2월 14일 현재 한국의 30일 평균 무위험이자율은 3.4129%입니다. 어느 기업 A의 30일 평균 주식 수익률이 7.4%였다고 할 때, 투자자가 이 기업 A에 기대하는 요구수익률은 얼마입니까?(베타는 1.2로 가정합니다)

#### Assignment 1-2.

- 위 기업의 경우에서 무위험이자율을 2.5%라고 가정할 때의 요구수익률을 계산하세요

### Assignment 2. 

Gordon의 주식 가치 결정 모델에서 적정 주가는 다음과 같이 정의됩니다

$$P = \frac{D(1+g)}{k-g}$$

여기서 $k$는 요구수익률, $D$는 배당금, $g$는 배당 증가율, $P$는 기업의 적정 주가를 의미합니다

- 위 산식을 함수로 작성하세요

- 과제 1번의 요구수익률을 가정하고, 기존의 기업 배당금은 1주당 1천원, 배당 증가율이 3%라면 이 기업의 적정 주가는 얼마입니까?